Fix missing icons

caksoylar opened the pull request:
Since the previous fixed path `images/icon.ico` stopped working for the store distros, modified the installation script to use the executable path for the icon. Since executable path is not the same as the distro name we get from the registry, read it from AppxManifest.xml as suggested by @Biswa96 in https://github.com/mintty/wsltty/issues/110#issuecomment-395999015.

It works on my machine but could use more testing. For example I had changed the owner of `C:Program Files\WindowsApps` to my user so I can open it in explorer -- there is a chance the path may not be readable by the script by default.

caksoylar commented:
@mintty Any comments on this approach? This goes through a powershell call in config-distros.sh to get the executable path and sets the icon using that.

Biswa96 commented:
There are some conflicts. Continue the discussion in your issue.

caksoylar commented:
What conflicts exactly?

mintty commented:
@caksoylar, sorry for the pending feedback, I lost awareness of this issue. Actually, there is a different approach from @Biswa96 for the same problem. I suggest you negotiate the best approach with each other, if possible, and then I'll take a look, as my capacity to deal with this is currently limited, sorry.

About a patch, please do not exchange TABs with spaces, so the diff doesn't get cluttered with distracting entries like
-  "")	# WSL default installation
+  "")  # WSL default installation

mintty commented:
Actually, it's not quite the same problem. The icon picked in the config script is for the shortcuts and start menu entries, while the icon determined by mintty itself is used for non-shortcut invocation.
But the algorithm to find the proper icon should be clarified first and should be the same in both cases.

mintty commented:
Actually, it's not quite the same problem. The icon picked in the config script is for the shortcuts and start menu entries, while the icon determined by mintty itself is used for non-shortcut invocation.
But the algorithm to find the proper icon should be clarified first and should be the same in both cases.

mintty commented:
Just installed Kali Linux.
@caksoylar, as you mentioned in https://github.com/mintty/wsltty/issues/110#issuecomment-395997596,
the launcher name is not kali-linux.exe as would be expected. I assume your script picks out the actual launcher name (kali.exe)? (Not yet tested.) However, the icon it contains is not the icon used for the Kali Linux start menu shortcut it installs... This is all too insane...

mintty commented:
In the case of Kali Linux, the entry you look up is "kali.exe"; why is this stripping needed?
> executable="${executable%"${executable##*[![:space:]]}"}"

mintty commented:
Also, I'd appreciate the previous choice as a fallback, with the two options selected by
> if [ -r "$ProgramW6432/WindowsApps/$instdir/images/icon.ico" ]

instead of always falling back to wsl.ico directly.

caksoylar commented:
Hi @mintty, thanks for the comments and the explanation about mintty shortcut invocation -- I wasn't aware that that was also a problem.

> In the case of Kali Linux, the entry you look up is "kali.exe"; why is this stripping needed?

I haven't been able to figure out exactly why, but when I do the powershell call and store the result in $executable, it contains a trailing newline, i.e. it is `kali-linux.exe\n` in this case. That in turn breaks the rest of the script. It is a bit ugly but that stripping works to get rid of the trailing newline.

> About a patch, please do not exchange TABs with spaces, so the diff doesn't get cluttered with distracting entries like

You are right, the script had mixed tabs and spaces so I had corrected them but I reverted the changes.

> Also, I'd appreciate the previous choice as a fallback

I also added that as a secondary check in the added commits.

> However, the icon it contains is not the icon used for the Kali Linux start menu shortcut it installs... This is all too insane...

I hadn't noticed that, that's really insane 😐 I assume that shortcut uses one of the assets under `$ProgramW6432/WindowsApps/$instdir/Assets/`. There are some related items with links to files under Assets in AppxManifest.xml, so we could extract one of those instead of the executable name?

mintty commented:
How do you test this? If I run Powershell from the script, using the cygwin shell (and even from within a Windows console), the Powershell command is run in a newly opened console window and its output is lost.

For slightly simplified quote escaping, I'd suggest:
`    	psh_cmd='([xml]$(Get-Content "'"$manifest"'")).Package.Applications.Application.Executable'`

> There are some related items with links to files under Assets in AppxManifest.xml, so we could extract one of those instead of the executable name?

I'm afraid for shortcut setup, the Windows icon format is needed (.ico or a resource in a .exe), so .png will not work.

mintty commented:
Maybe the XML extraction can be achieved with plain shell means instead of Powershell.
Using `sed` would be fine, but would also pull in library dependencies. Something along this line might work:
```
while read line
do
  case "$line" in
    *"<Application"*"Executable="*) echo $line;;
  esac
done < "$manifest"
```
Put the output in a shell variable:
```
exeline=`while ...`
```
and continue with some shell string magic...

mintty commented:
... but we can still use the icon we find in the launcher, even if it's different;
that's not our biggest problem

mintty commented:
Released 3.0.2.
